File: jAutoInvoice/src/java/net/sf/jautoinvoice/server/model/LinhaGasto.java
/*
 * LinhaGasto.java
 *
 * This file is part of jAutoInvoice, http://sourceforge.net/p/jautoinvoice
 *
 * Copyright (C) 2011 jAutoInvoice
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package net.sf.jautoinvoice.server.model;

import java.io.Serializable;

/**
 * 
 * @since 1.0
 */
public final class LinhaGasto implements Serializable {

    private String id;
    private Material material;
    private double quantidade;
    private double precoUnitario;
    private LinhaServico linhaServico;

    public LinhaGasto() {
        //DO NOTHING
    }

    public LinhaGasto(Material material, double quantidade, double precoUnitario,
            LinhaServico linhaServico) {
        this.material = material;
        this.quantidade = quantidade;
        this.precoUnitario = precoUnitario;
        this.linhaServico = linhaServico;
    }

    public LinhaServico getLinhaServico() {
        return linhaServico;
    }

    public void setLinhaServico(LinhaServico linhaServico) {
        this.linhaServico = linhaServico;
    }

    public Material getMaterial() {
        return material;
    }

    public void setMaterial(Material material) {
        this.material = material;
    }

    public double getPrecoUnitario() {
        return precoUnitario;
    }

    public void setPrecoUnitario(double precoUnitario) {
        this.precoUnitario = precoUnitario;
    }

    public double getQuantidade() {
        return quantidade;
    }

    public void setQuantidade(double quantidade) {
        this.quantidade = quantidade;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final LinhaGasto other = (LinhaGasto) obj;

        if (this.id != null && this.id.equals(other.id)) {
            return true;
        }

        if (this.material != other.material && (this.material == null || !this.material.equals(other.material))) {
            return false;
        }
        if (this.quantidade != other.quantidade) {
            return false;
        }
        if (this.precoUnitario != other.precoUnitario) {
            return false;
        }
        if (this.linhaServico != other.linhaServico && (this.linhaServico == null || !this.linhaServico.equals(other.linhaServico))) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 37 * hash + (this.id != null ? this.id.hashCode() : 0);
        hash = 37 * hash + (this.material != null ? this.material.hashCode() : 0);
        hash = 37 * hash + (int) this.quantidade;
        hash = 37 * hash + (int) this.precoUnitario;
        hash = 37 * hash + (this.linhaServico != null ? this.linhaServico.hashCode() : 0);
        return hash;
    }
}

File: jai/models/LinhaGasto.php
<?php

/* This file is part of jAutoInvoice, a car workshop management software.
 * 
 * Copyright (c) 2012 - 2014, Sérgio Lopes.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * @property int $idLinhaGasto
 * @property int $quantidade
 * @property float $precoUnitario
 * @property int $activo
 * 
 * @property int $idMaterial
 * @property int $idIVA
 * @property int $idLinhaServico
 *
 * @property Material $material
 * @property IVA $iva
 * @property LinhaServico $linhaServico
 */
class LinhaGasto extends CActiveRecord {

    /**
     * @return LinhaGasto
     */
    public static function model($className = __CLASS__) {
        return parent::model($className);
    }

    public function tableName() {
        return '{{LinhaGasto}}';
    }

    public function rules() {
        return array(
            array('quantidade, precoUnitario, idMaterial, idIVA, idLinhaServico', 'required'),
            array('quantidade, idLinhaServico, idIVA, idMaterial', 'numerical', 'integerOnly' => true),
            array('precoUnitario', 'numerical'),
            // search
            array('quantidade, idMaterial, idIVA, idLinhaServico', 'safe', 'on' => 'search'),
        );
    }

    public function relations() {
        return array(
            'linhaSercivo' => array(self::BELONGS_TO, 'LinhaServico', 'idLinhaServico'),
            'material' => array(self::BELONGS_TO, 'Material', 'idMaterial'),
            'iva' => array(self::BELONGS_TO, 'IVA', 'idIVA'),
        );
    }

    public function attributeLabels() {
        return array(
            'idLinhaGasto' => 'ID',
            'quantidade' => 'Quantidade',
            'precoUnitario' => 'Preço Uni.',
            'idMaterial' => 'Material',
            'idIVA' => 'IVA',
            'idLinhaServico' => 'Serviço',
        );
    }

    /**
     * @return CActiveDataProvider
     */
    public function search() {
        $criteria = new CDbCriteria();

        $criteria->compare('quantidade', $this->quantidade);
        $criteria->compare('precoUnitario', $this->precoUnitario);

        $criteria->compare('idMaterial', $this->idMaterial);
        $criteria->compare('idIVA', $this->idIVA);
        $criteria->compare('idLinhaServico', $this->idLinhaServico);

        $criteria->compare('activo', 1);

        return new CActiveDataProvider($this, array(
            'criteria' => $criteria,
            'sort' => array(
                'defaultOrder' => 'idLinhaGasto'
        )));
    }

}
